### PR TITLE
05 images

### PR DESCRIPTION
**Pull request by victorbtgn**
С тестами чёт пока не заходит(

### CODE REVIEW

File: api/auth/auth.controller.js [hunk 1]
```diff
@@ -44,24 +38,30 @@ const registerUserController = async (req, res, next) => {
   try {
     const { body } = req;
     const hashedPassword = await bcrypt.hash(body.password, +process.env.SALT);
+
     const user = await User.registerUser({
       ...body,
       password: hashedPassword,
     });
+
+    if (!user) return res.status(500).json({ message: "Internal Server Error" });
```
Comment (odykyi) [inline api/auth/auth.controller.js]: not 500

use 404 user not found

File: api/auth/auth.controller.js [hunk 2]
```diff
@@ -111,10 +112,26 @@ const logoutUserController = async (req, res, next) => {
   }
 };
 
+const uploadAvatarController = async (req, res, next) => {
+  try {
+    const user = await User.findUserById(req.user._id);
+    const oldUserAvatarName = await user.avatarURL.split("/")[4];
+    if (oldUserAvatarName) {
+      await fs.unlink(`public/images/${oldUserAvatarName}`);
```
Comment (odykyi) [inline api/auth/auth.controller.js]: move `public/images/${oldUserAvatarName}` to config

File: api/auth/auth.model.js [hunk 3]
```diff
@@ -4,22 +4,17 @@ const userSchema = new Schema({
...
     type: String,
     enum: ["free", "pro", "premium"],
     default: "free",
   },
-  token: {
+  token: String,
+  avatarURL: {
     type: String,
-    default: "",
+    // default: "http://localhost:8000/images/default-avatar.jpg",
```
Comment (odykyi) [inline api/auth/auth.model.js]: remove comments

File: middlewares/fileUploader.middleware.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+const multer = require("multer");
+
+const avatarUploader = () => {
+  const storage = multer.diskStorage({
+    destination: function (req, file, cb) {
+      cb(null, "tmp");
+    },
+    filename: function (req, file, cb) {
+      const fileType = file.mimetype.split("/")[1];
+      if (fileType !== "jpg" && fileType !== "jpeg" && fileType !== "png") return cb(new Error("File must be a photo"));
```
Comment (odykyi) [inline middlewares/fileUploader.middleware.js]: move all file formats to config as an array 
and check in array includes fileType

File: services/createDefaultAvatar.service.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  await Avatar.builder(Avatar.Image.circleMask(Avatar.Image.identicon()), 80, 80, {
+    cache: Avatar.Cache.folder("./tmp"),
+  }).create(userId);
+
+  return await minifyAvatar();
+};
+
+const minifyAvatar = async () => {
+  try {
+    const file = await imagemin(["./tmp/"], {
```
Comment (odykyi) [inline services/createDefaultAvatar.service.js]: move path to config

File: services/createDefaultAvatar.service.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+const fs = require("fs").promises;
+
+const Avatar = require("avatar-builder");
+const imagemin = require("imagemin");
+const imageminJpegtran = require("imagemin-jpegtran");
+
+const createDefaultAvatar = async (userId) => {
+  await Avatar.builder(Avatar.Image.circleMask(Avatar.Image.identicon()), 80, 80, {
+    cache: Avatar.Cache.folder("./tmp"),
```
Comment (odykyi) [inline services/createDefaultAvatar.service.js]: move path to config

File: services/createDefaultAvatar.service.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  }).create(userId);
+
+  return await minifyAvatar();
+};
+
+const minifyAvatar = async () => {
+  try {
+    const file = await imagemin(["./tmp/"], {
+      destination: "public/images",
```
Comment (odykyi) [inline services/createDefaultAvatar.service.js]: move path to config

File: services/createDefaultAvatar.service.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      plugins: [imageminJpegtran()],
+    });
+
+    await fs.unlink(file[0].sourcePath);
+
+    const destinationPath = await file[0].destinationPath.split("\\")[2];
+
+    return `http://localhost:8000/images/${destinationPath}`;
```
Comment (odykyi) [inline services/createDefaultAvatar.service.js]: move path to config host and port

File: test/testDB.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  });
+});
+
+const nonToken = () => "";
+
+it("Check auth middleware without token", (done) => {
+  const [req, res, next] = argForMiddleware();
+  req.get = nonToken;
+
+  checkAuthTokenMiddleware(req, res, next).then(() => {
```
Comment (odykyi) [inline test/testDB.js]: remove then

you should use await

File: test/testDB.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+});
+
+const invalidToken = () =>
+  "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVmNzRlYmQ5MGMwOGIzMzU5ODlkNzVjOCIsImlhdCI6MTYwMTY1NDAyMH0.bSaMyhIWYzjnx8aVIIDh8QwGLxY-UhmR4pjUT_nmanG";
+
+it("Check auth middleware with invalid token", (done) => {
+  const [req, res, next] = argForMiddleware();
+  req.get = invalidToken;
+
+  checkAuthTokenMiddleware(req, res, next).then(() => {
```
Comment (odykyi) [inline test/testDB.js]: remove then

you should use await

File: test/testDB.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+const validToken = () =>
+  "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVmNzRiNTE1ZTQ4ODRkMmM1NDcwMzA3NyIsImlhdCI6MTYwMTQ4NDEyOX0.YK-fzBeDVHckGo7NzuQTRxD1_9A2qjBViZqzabGqTlk";
+
+it("Check auth middleware with valid token", (done) => {
+  const [req, res, next] = argForMiddleware();
+  req.get = validToken;
+
+  checkAuthTokenMiddleware(req, res, next).then(() => {
```
Comment (odykyi) [inline test/testDB.js]: remove then

you should use await

**Review by odykyi — CHANGES_REQUESTED** [review]
comments

File: test/testDB.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+      .get('/current')
+      .set('Authorization', token)
+      .expect(200, done)
+      // .end((err, res) => {
+      //   console.log('ERR: ', err);
+      //   res.status.should.equal(200)
+      //   done()
+      // })
+
+    // request(app)
```
Comment (odykyi) [inline test/testDB.js]: why tests commented?

File: test/testDB.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  const token = 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVmN2Y0MTkxNWVmZDgxMjE0YzRlNjk5YyIsImlhdCI6MTYwMjE3NTQxOH0.9Dz4kOJ6hLT8R8gQGhNEfvSk52B_AQhajO1-q9G5W38';
+  const app = express();
+  app.use(express.urlencoded({ extended: false }))
+  app.use("/auth", authRouter);
+
+  it('Check middleware with invalid token',  (done) => {
+    request(app)
+      .get('/current')
+      .set('Authorization', token)
+      .expect(200, done)
```
Comment (odykyi) [inline test/testDB.js]: Check middleware with invalid token should be 200 ?

File: test/testDB.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    };
+  
+    return [req, res, next];
+  };
+  
+  const validToken = () =>
+    "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVmNzRiNTE1ZTQ4ODRkMmM1NDcwMzA3NyIsImlhdCI6MTYwMTQ4NDEyOX0.YK-fzBeDVHckGo7NzuQTRxD1_9A2qjBViZqzabGqTlk";
+  
+  it("Check auth middleware with valid token", async () => {
+    return new Promise(async (resolve, reject) => {
```
Comment (odykyi) [inline test/testDB.js]: is it possible to write code without new Promise ?

**Review by odykyi — CHANGES_REQUESTED** [review]
comments

File: test/testDB.js (hunk 13 above)
Comment (victorbtgn) [reply test/testDB.js]: no, i try 200 and 401, but always get 404, i don't understand why. i forgot remove 200

File: test/testDB.js (hunk 12 above)
Comment (victorbtgn) [reply test/testDB.js]: i can't end this test

File: test/testDB.js (hunk 14 above)
Comment (victorbtgn) [reply test/testDB.js]: is it a question or a statement ?

Comment (victorbtgn) [reply test/testDB.js]: yes, this is possible

**Review by odykyi — APPROVED** [review]

**Pull request merged**